File: swim-system-java/swim-core-java/swim.collections/src/main/java/swim/collections/STreePage.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.collections;

import java.util.Map;
import swim.util.Cursor;

public abstract class STreePage<T> {
  public abstract boolean isEmpty();

  public abstract int size();

  public abstract int arity();

  public abstract boolean contains(Object value);

  public abstract int indexOf(Object value);

  public abstract int lastIndexOf(Object value);

  public abstract T get(int index);

  public abstract Map.Entry<Object, T> getEntry(int index);

  public abstract STreePage<T> updated(int index, T newValue, STreeContext<T> tree);

  public abstract STreePage<T> inserted(int index, T newValue, Object id, STreeContext<T> tree);

  public STreePage<T> appended(T newValue, Object id, STreeContext<T> tree) {
    return inserted(size(), newValue, id, tree);
  }

  public STreePage<T> prepended(T newValue, Object id, STreeContext<T> tree) {
    return inserted(0, newValue, id, tree);
  }

  public abstract STreePage<T> removed(int index, STreeContext<T> tree);

  public abstract STreePage<T> removed(Object value, STreeContext<T> tree);

  public abstract STreePage<T> drop(int lower, STreeContext<T> tree);

  public abstract STreePage<T> take(int upper, STreeContext<T> tree);

  public abstract STreePage<T> balanced(STreeContext<T> tree);

  public abstract STreePage<T> split(int index);

  public abstract STreePage<T> splitLeft(int index);

  public abstract STreePage<T> splitRight(int index);

  public abstract void copyToArray(Object[] array, int offset);

  public Cursor<T> iterator() {
    return Cursor.values(this.entryIterator());
  }

  public Cursor<Object> keyIterator() {
    return Cursor.keys(this.entryIterator());
  }

  public abstract Cursor<Map.Entry<Object, T>> entryIterator();

  public Cursor<T> reverseIterator() {
    return Cursor.values(this.reverseEntryIterator());
  }

  public Cursor<Object> reverseKeyIterator() {
    return Cursor.keys(this.reverseEntryIterator());
  }

  public abstract Cursor<Map.Entry<Object, T>> reverseEntryIterator();

  public static <T> STreePage<T> empty() {
    return STreeLeaf.empty();
  }
}

File: swim-system-js/swim-core-js/@swim/collections/main/STreePage.ts
// Copyright 2015-2020 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Cursor} from "@swim/util";
import {KeysCursor} from "./KeysCursor";
import {ValuesCursor} from "./ValuesCursor";
import {STreeContext} from "./STreeContext";
import {STree} from "./STree";
import {STreeLeaf} from "./STreeLeaf";

/** @hidden */
export abstract class STreePage<V = unknown, I = unknown> {
  abstract get arity(): number;

  abstract get size(): number;

  abstract isEmpty(): boolean;

  abstract get(index: number): V | undefined;

  abstract getEntry(index: number): [I, V] | undefined;

  abstract updated(index: number, newValue: V, tree: STreeContext<V, I>): STreePage<V, I>;

  abstract inserted(index: number, newValue: V, id: I | undefined, tree: STreeContext<V, I>): STreePage<V, I>;

  abstract removed(index: number, tree: STreeContext<V, I>): STreePage<V, I>;

  abstract drop(lower: number, tree: STreeContext<V, I>): STreePage<V, I>;

  abstract take(upper: number, tree: STreeContext<V, I>): STreePage<V, I>;

  abstract balanced(tree: STreeContext<V, I>): STreePage<V, I>;

  abstract split(index: number): STreePage<V, I>;

  abstract splitLeft(index: number): STreePage<V, I>;

  abstract splitRight(index: number): STreePage<V, I>;

  abstract forEach<T, S>(callback: (this: S,
                                    value: V,
                                    index: number,
                                    tree: STree<V, I>,
                                    id: I) => T | void,
                         thisArg: S,
                         offset: number,
                         tree: STree<V, I>): T | undefined;

  keys(): Cursor<I> {
    return new KeysCursor(this.entries());
  }

  values(): Cursor<V> {
    return new ValuesCursor(this.entries());
  }

  abstract entries(): Cursor<[I, V]>;

  reverseKeys(): Cursor<I> {
    return new KeysCursor(this.reverseEntries());
  }

  reverseValues(): Cursor<V> {
    return new ValuesCursor(this.reverseEntries());
  }

  abstract reverseEntries(): Cursor<[I, V]>;

  private static _empty?: STreeLeaf<unknown, unknown>;

  static empty<V, I>(): STreeLeaf<V, I> {
    if (STreePage._empty === void 0) {
      STreePage._empty = new STree.Leaf([]);
    }
    return STreePage._empty as STreeLeaf<V, I>;
  }
}
STree.Page = STreePage;
